lib/src/core/engine.dart

```diff
@@ -636,15 +736,40 @@ class Engine extends Disposable with EventsEmittable<EngineEvent> {
     }
   }
 
-  void _onDCMessage(rtc.RTCDataChannelMessage message) {
+  void _onDCMessage(rtc.RTCDataChannelMessage message) async {
     // always expect binary
     if (!message.isBinary) {
       logger.warning('Data message is not binary');
       return;
     }
-
     final dp = lk_models.DataPacket.fromBuffer(message.binary);
+    if (dp.whichValue() == lk_models.DataPacket_Value.encryptedPacket) {
+      if (_e2eeManager == null) {
+        logger.warning('Received encrypted packet but E2EE not set up');
+        return;
+      }
+      final decryptedData = await _e2eeManager?.handleEncryptedData(
+        data: Uint8List.fromList(dp.encryptedPacket.encryptedValue),
+        iv: Uint8List.fromList(dp.encryptedPacket.iv),
+        participantIdentity: dp.participantIdentity,
+        keyIndex: dp.encryptedPacket.keyIndex,
+      );
+      if (decryptedData == null) {
+        logger.warning('Failed to decrypt data packet');
+        return;
+      }
+
+      final decryptedPacketPayload =
+          lk_models.EncryptedPacketPayload.fromBuffer(decryptedData);
+      final newDp = asDataPacket(decryptedPacketPayload);
+
+      _emitDataPacket(newDp);
```

Comment:
we should add some encryption metadata here, I opted to add the `encryptionType` for the data packet. 

This allows us also to error on data streams that have inconsistent encryption types during the lifetime of a stream